content/advent-2016/context-logging.md

```diff
@@ -0,0 +1,580 @@
+        rqId := uuid.NewRandom()
+        rqCtx := myappcontext.WithRqId(httpContext, rqId)
+        logger = myappctx.Logger(rqCtx)
+        logger.Info("handling /doit request", zap.String("for", user))
+        somelbrary.DoSomethingAwesome(rqCtx, ...)
+        logger.Info("finished handling /doit request")
+    }
+```
+
+Now, the requests will come through tagged!
+
+```
+2016/12/01 14:49:12 [INFO] handling /doit request rqId="aad6dcde..." for="Bert"
+2016/12/01 14:49:12 [INFO] handling /doit request rqId="7f5b859a..." for="Alex"
+2016/12/01 14:49:12 [INFO] did something awesome rqId="aad6dcde..." so="totally" param="cheese" wow="3"
+2016/12/01 14:49:12 [INFO] did something awesome rqId="7f5b859a..." so="so" param="param" wow="2"
+2016/12/01 14:49:12 [INFO] finished handling /doit request rqId="7f5b859a..."
+2016/12/01 14:49:12 [INFO] finished handling /doit request rqId="7f5b859a..."
+```
+
+It's now possible to see which line was handled by each request
+handler, even though they are arriving in the log interleaved with
+each other.
+
+## Wait, so what is `context.WithValue` doing?  Does it return a copy?!
+
+A good conceptual model of context's value storage is a map from
+anything (`interface{}`) to anything else (`interface{}`).
+`context.WithValue` returns a new context object which will return the
+value you poked in when you look up via the key you poked it into
+using `.Value`.
+
+Similarly to a map with a key `interface{}`, lookup in a context is a
+comparison where the value **and the type** of the key are matched
+exactly.  Using a custom `int` type fits this bill perfectly, and
+means that the value lookup is very fast: normally just comparing one
+pointer (to the type structure) and one value: both single machine
+words.
+
+Each context object is also immutable, although the values in it don't
+necessarily have to be.  The way to update maps with immutable objects
+in functional languages like Erlang is to create a new one with the
+new key added, and this is something like what `context.WithValue`
+does.  It's actually a singly linked list, and each link only holds a
+single key.  It looks it up, and if it doesn't find it, it passes it
+to the parent.
+
+What this immutability means is that it can be safely shared - across
+goroutines, functions etc, and the state is only ever additive.
+Important information that you add is rolled back as the functions
+return.  It also makes it very cacheline friendly for decent
+multi-core performance.
+
+## So, what happens if the key isn't set?  Nil Pointer Panic?
+
+Values not being set in the context is explicitly encouraged. If you
+try to fetch from the context and the key is not set, you'll get `nil`
+back instead.
+
+Did I say "encouraged"?  What I actually mean is, if your function
+doesn't work at all unless magic values are poked into the context
+first, you are almost certainly abusing `context`.
+
+You should always try to cast the value back to the type you expected
```

Comment:
technically this is called a "type assertion", not "casting"
https://golang.org/ref/spec#Type_assertions